Python/bltinmodule.c

```diff
@@ -157,6 +157,13 @@ builtin___build_class__(PyObject *self, PyObject **args, Py_ssize_t nargs,
         Py_DECREF(bases);
         return NULL;
     }
+    if (!PyMapping_Check(ns)) {
+        PyErr_Format(PyExc_TypeError,
+                     "%.200s.__prepare__() must return a mapping, not %.200s",
+                     isclass ? ((PyTypeObject *)meta)->tp_name : "<metaclass>",
```

Comment:
Why this code is as it is?

Reply:
I thought that printing the name of the problematic metaclass would be helpful to the user.
Now ISTM that I should have checked the ``__name__`` attribute.
Do you think I should add that? Or should we just not include the name of the metaclass in the error message?

Reply:
Why not use `((PyTypeObject *)meta)->tp_name` in all cases?

Reply:
IIUC, if `meta` is not a type object, then `((PyTypeObject *)meta)->tp_name` might contain arbitrary data, according to the type of `meta`, which might cause a crash (as indeed happens when running my second test).

Reply:
In what circumstances `meta` is not a type object?

Reply:
This code is an example:
```
class BadMetaclass:
    def __prepare__(*args):
        pass

class Bar(metaclass=BadMetaclass()):
    pass
```
I don't know about usecases for that (though I am certainly not a metaclasses expert), but IIUC, the C code explicitly handles such cases.

Reply:
Maybe use `meta->ob_type->tp_name` in this case?

Reply:
When metaclass isn't a type, it can be an arbitrary callable, so I think it's reasonable to just hardcode in `<metaclass>` for that case.